Initial commit based on preliminary work

david-bouyssie opened the pull request:
Code taken from these two existing repos:
* https://github.com/pinaraltiner/proteomics-rs/
* https://github.com/lazear/sage/

david-bouyssie commented:
[proteomics-rs/](https://github.com/pinaraltiner/proteomics-rs/) corresponds to a repo I contributed to with Pinar Altiner (PhD in our lab).

lazear commented on mzcore-rs/src/msms/annotator.rs:
```diff
@@ -0,0 +1,300 @@
+    pub aa_position: u16, // AA position in amino acid sequence (starts at 1)
+}
+
+pub fn annotate_spectrum(spectrum_peaks: &Vec<[f64;2]>, frag_table: &FragmentationTable, mz_error_tol: f64) -> Vec<MatchedPeak> {
```
Nit: in general, it is always preferable to take a slice instead of a vec by reference (`&[[f64; 2]]`). And, once a peak representation is settled on, would be better to use that of course.

lazear reviewed (commented):
I don't want to provide too many specific comments in the spirit of just making progress. 

Some general comments: 
- There are lots of functions that could be methods instead - having associated methods helps *a lot* with API surface exploration and IDE hints. For example, there are lots of functions that take ` aa_table: &AminoAcidTable` as an argument... perhaps they could all be methods of `AminoAcidTable` instead?
- In a similar vein, there are benefits to having a newtype representation of AminoAcid - can have composition, pka, etc be associated methods. This also enables parsing a string into a `Vec<AminoAcid>` or something. I am a strong believer in using the type system to enforce invariants/only allow legal data representations (impossible to have an invalid string of AminoAcids, etc).

david-bouyssie commented:
Thank you for your quick review @lazear 
Nice suggestions.

I'll try to introduce these changes tomorrow.

david-bouyssie commented on mzcore-rs/src/chemistry/mass_calc.rs:
```diff
@@ -0,0 +1,67 @@
+    calc_aa_seq_mass_from_chars(char_iter, aa_table, mono_mass)
+}
+
+pub fn calc_aa_seq_mass_from_chars(aa_as_chars: std::slice::Iter<char>, aa_table: &AminoAcidTable, mono_mass: bool) -> Result<f64> {
```
So it should be renamed `calc_aa_seq_mass()` with this API?

david-bouyssie replied:
@lazear 
What is the issue about this type?
https://github.com/rusteomics/mzcore/blob/57f90663a5c19bfcd2acec53f2f58931b48bfb59/mzcore-rs/src/chemistry/model.rs#L49

david-bouyssie replied:
>This also enables parsing a string into a Vec<AminoAcid>
I think there is a trade-off to find between safety and simplicity. For many applications dealing with Iter<char> is far from enough and can provide some performance benefit on large volume of data.
I think we should support various APIs for different purposes.

david-bouyssie replied:
>There are lots of functions that could be methods instead - having associated methods helps a lot with API surface exploration and IDE hints. For example, there are lots of functions that take  aa_table: &AminoAcidTable as an argument... perhaps they could all be methods of AminoAcidTable instead?

Actually, could you please be more specific? Which precise functions would you to migrate to methods?

lazear replied:
> @lazear What is the issue about this type?
> 
> https://github.com/rusteomics/mzcore/blob/57f90663a5c19bfcd2acec53f2f58931b48bfb59/mzcore-rs/src/chemistry/model.rs#L49

It is quite large (~64 bytes per AminoAcid?)! But it's along the lines of what I was thinking. 

> Actually, could you please be more specific? Which precise functions would you to migrate to methods?

There are still some `compute_frag...` methods, but the other couple are now under a trait `SeqStrMassCalc`.

What is the exact purpose of needing an `AminoAcidTable` for so many functions was more what I was getting at - why not just represent amino acids more simply? How frequently will users actually need to define custom amino acids (which seems to be the purpose for the AminoAcidTable struct)? <1% of the time? I think this kind of dovetails with my feedback on the initial rusteomics stuff, which is that some of it seems over-engineered - IMO, the best approach is to implement things for the 99% use-case and get rusteomics actually used in some projects so that the correct parts can be over-engineered/flexible.

Something along the following lines seems much simpler to me...
```rs
trait AminoAcid {
    fn mass(&self) -> f32;
    fn three_letter(&self) -> &'static str;
}

impl AminoAcid for u8 {
    fn mass(&self) -> f32 {
        12.0
    }
    
    fn three_letter(&self) -> &'static str {
        match self {
            b'A' => "Ala",
            b'C' => "Cys",
            _ => unimplemented!()
        }
    }
}

trait Peptide {
    fn mass(&self) -> f32;
}

impl Peptide for &str {
    fn mass(&self) -> f32 {
        self.as_bytes().iter().copied().map(|aa| aa.mass()).sum()
    }
}

fn some_generic_function<A: AminoAcid>(aa: A) -> f32 {
    aa.mass() + 42.0
}

fn main() {
    assert_eq!("PEPTIDE".mass(), 12.0 * 7.0);
}
```

Statically define the standard amino acids, use a single byte for their representation, and allow strings (can even make a newtype ValidatedPeptide or w/e). If those 1% of users need a more complex representation, then they can define their own datatype that implements `AminoAcid` or `Peptide`, etc.

david-bouyssie replied:
Ok I get your points.
It's more about have a standard set of methods around primitive types.
It makes sense.

I think allowing to switch from lightweight to complex repr also make sense.
For now end-users can use this static struct for this prupose:
[pub struct AminoAcidTable ](https://github.com/rusteomics/mzcore/blob/57f90663a5c19bfcd2acec53f2f58931b48bfb59/mzcore-rs/src/chemistry/table.rs#L89)

We could imagine an API like this:
```
trait AminoAcid {
    fn three_letter(&self) -> &'static str;
    fn mono_mass(&self) -> f64;
    fn average_mass(&self) -> f64;
    fn full_definition(&self) -> AminoAcidResidue;
}

impl AminoAcid for u8 {

    fn three_letter(&self) -> Option<&'static str> {
        match self {
            b'A' => Some("Ala"),
            b'C' => Some("Cys"),
            _ => None
        }
    }

    fn mono_mass(&self) -> Option<f64> {
        self.full_definition().map(|aa| aa.mono_mass())
    }
    
    fn average_mass(&self) -> Option<f64> {
        self.full_definition().map(|aa| aa.average_mass())
    }

    fn full_definition(&self) -> Option<AminoAcidResidue> {
        &*PROTEINOGENIC_AMINO_ACID_TABLE.aa_by_code_1.get(self as char)
    }
}
```

I designed the amino acid table to be somewhat flexible (non entirely static), but maybe its a wrong decision.
I wanted to have a kind of configurable amino acid definition.

On way to keep some kind of configurability would be to change slightly the API:

```
trait AminoAcid {
    fn three_letter(&self) -> &'static str;
    fn mono_mass(&self, aa_table: &AminoAcidTable) -> f64;
    fn average_mass(&self, aa_table: &AminoAcidTable) -> f64;
    fn full_definition(&self, aa_table: &AminoAcidTable) -> AminoAcidResidue;
}

impl AminoAcid for u8 {
    fn three_letter(&self) -> Option<&'static str> {
        match self {
            b'A' => Some("Ala"),
            b'C' => Some("Cys"),
            _ => None
        }
    }

    fn mono_mass(&self, aa_table: &AminoAcidTable) -> Option<f64> {
        self.full_definition(aa_table).map(|aa| aa.mono_mass())
    }
    
    fn average_mass(&self, aa_table: &AminoAcidTable) -> Option<f64> {
        self.full_definition(aa_table).map(|aa| aa.average_mass())
    }
    
    fn full_definition(&self, aa_table: &AminoAcidTable) -> Option<AminoAcidResidue> {
        aa_table.aa_by_code_1.get(self as char)
    }
}
```

But this is less straightforward...
To conclude, that's an interesting design question: shall allow non-static definitions of amino acids?

lazear replied:
If the purpose **wasn't** to allow non-static definitions of amino acids, I am really wondering **why** there is a `Table` struct. There is no good reason for passing it around if users are **never** going to construct different versions. The same goes for all of the other `Table` types - are users going to be defining their own, newly discovered `Atom`s? It just seems like strange API design - this can all be hidden from the user, why do they need to care what our internal representation of an `Atom` or `AminoAcid` is? Just define a lookup table (e.g. the array of monoisotopic masses, and be done with it - or attach the lookup tables inside of methods defined on a trait - or just use `STANDARD_AMINO_ACID_TABLE` internally without exposing it on the API surface)

Passing around `AminoAcidTable` (or PtmTable, AtomTable, etc) by ref implies that there is a dependency of any peptide on the **exact** table from which it was created - but this isn't encoded in the type system (ideally it would be, but this would require scattering lifetimes throughout the codebase everywhere, see below). What happens if a user changes the `AminoAcidTable` at some point in the code, and this dependency isn't part of the type system (which would prevent such changes from happening, given Rust's ownership rules). 

```rs
pub struct Peptide<'a> {
    sequence: String,
    mono_mass: f64,
    average_mass: f64,
    table: &'a AminoAcidTable, // Must be defined here for invariants to hold
}
```

^ -- To be clear, I am not advocating for the above struct, I think this would be an extremely unfortunate design decision.

In your example above, why not just do the following:
```rs

impl AminoAcid for u8 {
    fn three_letter(&self) -> &'static str {
        match self {
            b'A' => "Ala",
            b'C' => "Cys",
            _ => panic!("invalid amino acid!")
        }
    }

   fn is_valid(&self) -> bool {
      // Check if any other functions will panic.
   }

    ...

    fn average_mass(&self) -> f64 {
        // Why would we ever need to look this up? We are __already__ defining all of the valid
        // amino acids ourselves.
        // self.full_definition(aa_table).map(|aa| aa.average_mass())
       match self {
            b'A' => 42.0,
            _ => panic!("invalid amino acid!"),
       }
    }
    
```

david-bouyssie replied:
Actually my idea was to allow non -static definitions of amino acids. For instance if someone wanted to use other available information about them (different masses, pKA). O think there is not always a consensus about AA pKA...
Regarding masses it's somewhat standardized.
I'm not against your proposed design, I just wanted to clarify the initial goal.

david-bouyssie replied:
>Passing around AminoAcidTable (or PtmTable, AtomTable, etc) by ref implies that there is a dependency of any peptide on the exact table from which it was created

I don't think this is actually necessary.  The sequence and modifications are what identifies the peptide. The masses and other values are just computed properties. We don't necessarily need to know how they were computed.

But again this is not an argument to defend the current design.

david-bouyssie replied:
Now if had to change the design I think I would:
1. try to stick to your lightweight trait/implem wrapping u8
2. still allow some kind of configurability, event if by default a given definition would be used. One option could be to define a global/static API to configure amino acid properties. I avoided that in the first intent because I don't like global configurations. Another option could be to do this at compile time maybe using macros and external files?

I guess what you suggest is to postpone the second point to later, to avoid over engineering. That's sound reasonable we just have to be sure there is a workaround,  and that the proposed design is not a dead end in terms of configuration possibilities.

lazear replied:
Point #2 is already covered by using traits:
```rs
fn some_generic_function<A: AminoAcid>(aa: A) -> f32 {
    aa.mass() + 42.0
}

struct SagePeptide {
    sequence: std::sync::Arc<[u8]>,
    modifications: Vec<f32>,
}

impl Peptide for SagePeptide {
    fn mass(&self) -> f32 {
        self.sequence
            .as_ref()
            .iter()
            .copied()
            .map(|aa| aa.mass())
            .sum::<f32>()
            + self.modifications.iter().copied().sum::<f32>()
    }
}

struct MyCustomAA {
    n_heavy_isotopes: usize,
    base_aa: u8,
}

impl AminoAcid for MyCustomAA {
    fn mass(&self) -> f32 {
        self.base_aa.mass() + self.n_heavy_isotopes as f32 * 1.05
    }

    fn three_letter(&self) -> &'static str {
        self.base_aa.three_letter()
    }
}

fn main() {
    dbg!(some_generic_function(MyCustomAA { base_aa: b'A', n_heavy_isotopes: 2}));
    dbg!(some_generic_function(b'A'));
}
```

The benefit to the approach above is that rusteomics can provide a strong default option (u8, String) for AA's and peptides. Users with specific needs can just define their own types, and implement the traits. All of the library functions are generic over the traits (e.g. agnostic to the underlying representation - it doesn't matter whether a peptide is a `String` or a `Arc<[u8]>`).

lazear replied:
I'm not saying that a `AminoAcid` and `Peptide` trait are strictly necessary or the correct abstraction (could cover them with other traits, like `Mass`, `HasPka`, whatever) - but I think it covers the need for extensibility with defaults, and also has the benefit of allowing interoperability with other Rust programs (which is, I asssume, the goal for rusteomics), without forcing them to use a specific memory representation (or look up amino acid masses by using a HashMap).

david-bouyssie replied:
That's sounds like a nice way to go :+1:

david-bouyssie commented on mzcore-rs/src/ms/spectrum.rs:
```diff
@@ -0,0 +1,67 @@
+
+// --- Similar to sage definitions --- //
+
+/// A charge-less peak at monoisotopic mass
```
Nice catch. I switched from mass to mz but I forgot to remove the comment.
I just wanted to introduce a simple Peak type.

lazear commented on mzcore-rs/src/ms/utils.rs:
```diff
@@ -118,41 +118,50 @@ pub fn mass_to_mz(mass: f64, charge: i32) -> f64 {
-            ))?;
+                "can't find amino acid '{}' in the provided table",
+                aa_as_char
+            )).unwrap(); // You might want to handle this error better.
```
Why `unwrap()` if the return type is `Result`? Doesn't this defeat the purpose of returning a `Result`?

david-bouyssie replied:
Agreed, solved in last commit.

david-bouyssie commented on mzcore-rs/src/ms/utils.rs:
```diff
@@ -181,7 +190,7 @@ where
 
     let right_idx = match slice[left_idx..].binary_search_by(|a| key(a, &high)) {
-        Ok(idx) | Err(idx) => {
+        core::result::Result::Ok(idx) | core::result::Result::Err(idx) => {
```
Fixed

douweschulte commented on mzcore-rs/src/msms/fragmentation.rs:
```diff
@@ -0,0 +1,367 @@
+    let pep_seq_as_chars = pep_seq.chars();
+
+    // --- This function contains ion series directions itself --- //
+    use FragmentIonSeriesDirection::*;
```
Nit: when used this way (import an enum as * and then use the names directly in a match) the error messages after renaming an enum variant will be quite bad. Indicating there is no such name in the full scope. While if you import an enum as a shorter name (eg `use FragmentIonSeriesDirection as FISD`) and then use this shorter name in the match statement (eg `FISD::FORWARD`) the error messages will be immediately clear. Also very nitpicky again, names for enum variant are normally camel case and not all caps.

david-bouyssie replied:
+1 regarding use FragmentIonSeriesDirection as FISD

for the all caps I thinking I wanted to have name conflict between NONE and core::option::Option::None.
But switching to FISD actually solves that.
Also NONE could be replaced by a better name.

david-bouyssie replied:
Actually, regarding  `use FragmentIonSeriesDirection::*`, I don't think this an issue when used in combination of enum matching.

douweschulte commented on mzcore-rs/src/chemistry/table.rs:
```diff
@@ -0,0 +1,510 @@
+                Isotope::new(1, 2.0141017778, 0.000115).unwrap(),
+                // Isotope(3, 3.016049277725, 0.0f)
+            ]).unwrap(),
+            // TODO: add the proton to the table ???
```
My way of handling this was adding an electron to the table and seeing a proton as a `hydrogen - electron` this also easily allowed me to handle other charged species of other elements and/or electrons directly.

david-bouyssie replied:
I don't get your point.
Electron and proton masses are defined as constants there:
https://github.com/rusteomics/mzcore/blob/350ca0621998d5962d85133d4404982612cca898/mzcore-rs/src/chemistry/constants.rs#L8

What would you like to change?

douweschulte replied:
Nothing then. I just read to TODO question if a proton should be added and reacted based on that.

david-bouyssie replied:
Ah ok. I get it now, thanks.

douweschulte reviewed (commented):
I gave it a cursory read as well and left some comments inline.

douweschulte commented on mzcore-rs/src/chemistry/constants.rs:
```diff
@@ -0,0 +1,90 @@
+    b'T', b'V', b'W', b'Y', b'U', b'O',
+];
+
+pub mod atom {
```
This to me looks like a nice option for an enum:
```rust
#[repr(u8)]
pub enum Element {
    C = b'C',
    O = b'O',
}
```
This same could be said about the amino acid mod above. This way it is impossible to write an incorrect Element when the argument for a function needs one, while the actual binary representation is exactly the same and to go from one to the other is as simple as a single cast `Element::C as u8`.

mobiusklein replied:
It is not possible to create a text mapping like that for elements whose names are longer than one character (e.g. Fe, Li, Na, Ca) which may be found as charge carriers on peptides or as part of modifications (or when a charge carrier is treated as a modification).

douweschulte replied:
That is true, for that reason I used an enum and provided a const function to give the symbol. (Although technically you could save the two bytes needed for unicode representation of these in a single `u16`)

david-bouyssie replied:
What about this?

```rust
mod element {
    pub const C: &'static str = "C";
    pub const H: &'static str = "H";
    pub const O: &'static str = "O";
    pub const N: &'static str = "N";
    pub const P: &'static str = "P";
    pub const S: &'static str = "S";
}

#[derive(Clone, Copy, Eq, PartialEq, Debug)]
pub enum Element {
    C,
    H,
    O,
    N,
    P,
    S,
}

impl Element {
    pub fn to_str(&self) -> &'static str {
        match self {
            Element::C => atom::C,
            Element::H => atom::H,
            Element::O => atom::O,
            Element::N => atom::N,
            Element::P => atom::P,
            Element::S => atom::S,
        }
    }
}

impl ToString for Element {
    fn to_string(&self) -> String {
        self.to_str().to_string()
    }
}
```

douweschulte replied:
Personally I would remove the `&str` constants, but I get why you have them in the first place. If you make the `to_str` function `const` you can just use that instead of the associated constants (except for using them as patterns in the match arm). Also I would never implement `ToString` I would rather opt for `std::fmt::Display` which gives you `ToString` 'for free' while also working with all the other print methods. And I would prefer an implementation of `TryFrom<&str> for Element` over the `new` function you use here, in that way more function will automatically work from the whole Rust ecosystem.

david-bouyssie replied:
What is the issue about non-exports constants?
+1 for std::fmt::Display instead of ToString (I figured it out in the meantime)
What about `impl FromStr for Element { fn from_str(value: &str) -> Result<Self, Self::Err> }` instead of `TryFrom<&str> for Element`?

douweschulte replied:
My issue personally is that introduces another link of C => "C", so you have to provide that link twice once fro the to_str and once for the constant. But as you used the constant in your parse method that issue is essentially solved.
For the trait `FromStr` is indeed preferred over `TryFrom<&str>`.

douweschulte replied:
Perfect, fully agreed. It should not change until someone manages to make `Uun`: `ununnonnium`.

douweschulte commented on mzcore-rs/src/chemistry/constants.rs:
```diff
@@ -0,0 +1,90 @@
+}
+
+// --- Sage definition --- //
+pub const MONOISOTOPIC_MASSES: [f32; 26] = [
```
Maybe this is a bit more fundamental point, I had to move away from direct lookup numbers for aminoacids thanks to some of the rules defined in ProForma. There are global modifications that replace all C12 by C13 (ProForma v2 4.6.1 - page 20) to handle this accurately you have to have an elemental composition of the aminoacids. Additionally working with everything in molecular formulas / elemental composition made it kind of nice for me to define constants, now it is just a constant molecular formula instead of a magic number. The downside is that you have to calculate the mass from this molecular formula, which takes some time. For me this has not been a problem yet, but I also have not thoroughly investigated yet so I cannot say how much of a percentage of the runtime this is for my usecases.

david-bouyssie replied:
I would favor the implementation of different APIs:
* one with built-in values (and lookup tables like this one) for simple cases
* another more complex API to solve issues like the one mentioned for ProForma (C12 vs C13). For this second case, we could provide atomic formula parsers, and mass calculators. I guess you already have that in rustyms?

douweschulte replied:
This is the reason I wrote my suggested trait to be generic over the output (mass or formula). And yes I have parser for both unimod and pro forma chemical formulas as well as a macro for unimod ones. Combined with all the mass calculators I needed. Although I would like to discuss a few points about them before directly pulling them in because I do not like a couple of small things about my implementation which you guys can maybe help to get rid of.

david-bouyssie replied:
Sounds great. Maybe we could setup a dedicated pro forma branch that would hold successive PRs.
But I guess we should wait for this first PR being merged.

douweschulte commented on mzcore-rs/src/chemistry/model.rs:
```diff
@@ -0,0 +1,293 @@
+
+#[derive(Clone, Copy, Eq, PartialEq, Debug)]
+pub enum PtmLocation {
+    PROT_N_TERM, // = Value("Protein N-term")
```
Rust formatting here suggests to use camelcase names, full uppercase is solely used for constants.

douweschulte commented on mzcore-rs/src/chemistry/model.rs:
```diff
@@ -0,0 +1,293 @@
+        use PtmLocation::*;
+
+        match self {
+            PROT_N_TERM => write!(f, "Protein N-term"),
```
Instead of importing `Self` here you can also use the `Self` shortcut and use that in the match statement. `Self` is the type this impl block is working with (`PtmLocation` in this case`) so the folllowing would be entirly valid:
```rust
write!(f, "{}",
match self {
   Self::ProteinNTerm => "Protein N-term",
  ...
})
```
This prevents the bad error messages I commented about last time, and is not that much longer to type.

douweschulte commented on mzcore-rs/src/chemistry/model.rs:
```diff
@@ -0,0 +1,293 @@
+
+#[derive(Clone, PartialEq, Debug)]
+struct AminoAcidPtm {
+    pub id: i64,
```
This struct is not at all set up to handle the width of PTMs that are needed for ProForma support. For reference [here is what I needed for nearly full pro forma support in rustyms](https://github.com/snijderlab/rustyms/blob/a9d9c8a83fbb866eeef34390513ebe3c5909d708/src/shared/modification.rs#L3).

douweschulte commented on mzcore-rs/src/chemistry/model.rs:
```diff
@@ -0,0 +1,293 @@
+
+#[derive(Clone, Default, PartialEq, Debug)]
+pub struct Peptide {
+    sequence: String,
```
To take advantage of the 'rust makes invalid states unrepresentable' property I would suggest to create something like the [`SequenceElement`](https://github.com/snijderlab/rustyms/blob/a9d9c8a83fbb866eeef34390513ebe3c5909d708/src/linear_peptide.rs#L471) (not sure about the name) struct I made in rustyms. This also has some advanced ProForma stuff, but in the basics it is just a combination of an `Aminoacid` (see comment above on how to make these guarenteed to be valid on compile time) joined with a list of possible modifications (yes a list, needed for ProForma). A peptide then should be really easy: just a list of this construct. (Although [the one in rustyms](https://github.com/snijderlab/rustyms/blob/a9d9c8a83fbb866eeef34390513ebe3c5909d708/src/linear_peptide.rs#L21) is slightly more fleshed out to support other use cases from ProForma as well.)

douweschulte commented on mzcore-rs/src/ms/utils.rs:
```diff
@@ -0,0 +1,226 @@
+    }
+}
+
+pub trait SeqCharsMassCalc {
```
The function type is exactly the same as `SeqStrMassCalc` so you could just reuse that trait.

douweschulte commented on mzcore-rs/src/msms/fragmentation.rs:
```diff
@@ -0,0 +1,367 @@
+
+
+pub fn compute_frag_table_with_mods(
+    pep_seq: &str,
```
I suggest to move this method to either use a `Peptide` or be a method on `Peptide` and I suggest the later, this makes it easier to discover the API surface and have nice hints pop up in your code editor.

douweschulte commented on mzcore-rs/src/msms/model.rs:
```diff
@@ -0,0 +1,370 @@
+
+#[allow(non_camel_case_types)]
+#[derive(Clone, Copy, Eq, PartialEq, Debug)] //, strum_macros::Display
+pub enum FragmentIonSeries {
```
If you want to reduce the number of enum options here I suggest to separate the `IonSeries` from the `NeutralLoss` that way the list of `IonSeries` reduces by 1/3 and potentially it can give you some other nice things as well, like being able to directly search for any y series ion and getting all neutral loss variants as well.

douweschulte commented on mzcore-rs/src/msms/model.rs:
```diff
@@ -0,0 +1,370 @@
+        c_NH3 => - WATER_MONO_MASS,
+        d => 0.0, // FIXME
+        v => 0.0, // FIXME
+        w => 0.0, // FIXME
```
These are based on the amino acid in question, so somewhat hard to calculate without that information. (Here)[https://github.com/snijderlab/rustyms/blob/a9d9c8a83fbb866eeef34390513ebe3c5909d708/src/aminoacids.rs#L105] is how that table is defined in rustyms.

douweschulte commented on mzcore-rs/src/msms/model.rs:
```diff
@@ -0,0 +1,370 @@
+    use FragmentIonSeriesDirection as SeriesDir;
+
+    let ion_series_direction = match ion_type {
+        a => SeriesDir::Forward,
```
clippy suggestion: you can merge multiple arms that have the same result:
```rust
a | b | c | ... => SeriesDir::Forward,
...
```

Also a style comment: I would pretty much always make a method that only has one argument and has a very logical connection to the type of that argument to be a method on the type, so here that could become `FragmentIonSeries::direction(&self)->FragmentIonSeriesDirection`

douweschulte commented on mzcore-rs/src/msms/model.rs:
```diff
@@ -0,0 +1,370 @@
+        c_p2 => SeriesDir::Forward,
+        c_H2O => SeriesDir::Forward,
+        c_NH3 => SeriesDir::Forward,
+        d => SeriesDir::Unspecified, // FIXME
```
d and v are in the same direction as a/b/c. w is on the same direction as x/y/z.
Also naming thing: would renaming the directions to N/Cterminal not make more sense?

douweschulte commented on mzcore-rs/src/msms/model.rs:
```diff
@@ -0,0 +1,370 @@
+        immonium => SeriesDir::Unspecified,
+    };
+
+    ion_series_direction
```
rust style thingy you can leave out the intermediate variable by removing the let statement and the semicolon at the end of the match statement. The variable name is not very helpful in this case and the resulting binary code is exactly the same.

douweschulte commented on mzcore-rs/src/msms/model.rs:
```diff
@@ -0,0 +1,370 @@
+            z => write!(f, "z"),
+            z_NH3 => write!(f, "z-NH3"),
+            z_H2O => write!(f, "z-H2O"),
+            z_dot => write!(f, "z."), // FIXME: is this the usual repr?
```
I used `·` 'middle dot' U+00B7, and because rust supports unicode names for variables i even used this [in the name for the series](https://github.com/snijderlab/rustyms/blob/a9d9c8a83fbb866eeef34390513ebe3c5909d708/src/fragment.rs#L241C6-L241C7).

david-bouyssie replied:
Rust complains with a warning that I was not able to hide/allow ("Non-ASCII characters").

douweschulte replied:
I do not know this warning. I have no allowed lints in my code that look something like this, it just works. Are you sure you are running a recent Rust version (`rustup update stable`)? Or maybe is it something in your editor?

david-bouyssie replied:
It was a parameter of my IDE, thanks!

douweschulte commented on mzcore-rs/src/msms/model.rs:
```diff
@@ -0,0 +1,370 @@
+
+#[derive(Clone, Copy, Eq, PartialEq, Debug)]
+pub enum FragmentType {
+    IMMONIUM,
```
Here I would suggest to use camelcase again.

douweschulte reviewed (commented):
I did a bit more thorough look through the code. A couple of comments could also be fit for a more general discussion so if no immediate resolve is found I will move them to a separate github discussion. Also I think many of them could be part of a general suggestions PR had this code been merged.

mobiusklein commented on mzcore-rs/src/ms/spectrum.rs:
```diff
@@ -0,0 +1,66 @@
+}
+
+pub trait HasSpectrumData {
+    fn get_mz_list(&self) -> &Vec<f64>;
```
Unless you intend to have a paired trait that is both mutable _and_ wants to be able to resize them, it's better for the trait to only ask for slices, not actual `Vec`. It's both more flexible and more idiomatic.

douweschulte commented:
I propose to merge this PR (unless @david-bouyssie still wants to work on it in this fashion) and have the comments above be part of a new PR by the commenter to have a bit better overview of what still needs to happen, be able to iterate faster, and be able to play around with fitting our own code inside the proposed framework. We do not have to have the full API stabilized in just this PR. We need to have two reviewers green-light this PR to make it mergable right? I do not think I have access to do that (yet?). Before merging @lazear also has to revoke his request for changes.

david-bouyssie commented:
> I propose to merge this PR (unless @david-bouyssie still wants to work on it in this fashion) and have the comments above be part of a new PR by the commenter to have a bit better overview of what still needs to happen, be able to iterate faster, and be able to play around with fitting our own code inside the proposed framework. We do not have to have the full API stabilized in just this PR. We need to have two reviewers green-light this PR to make it mergable right? I do not think I have access to do that (yet?). Before merging @lazear also has to revoke his request for changes.

I have several pending local changes. Some of them addresses issues mentioned int this PR. Some other changes are a proposal regarding the concerns raised in this inline discussion 
https://github.com/orgs/rusteomics/discussions/10
I can add them to this PR or to a separate one. It's up to you.

douweschulte commented:
I would suggest to include pending changes that address issues raised in this PR in this PR. But to keep the proposal changes outside, if you already mixed both in your history it is maybe easier to just do all of these in a new PR.

david-bouyssie commented:
Yes it's a bit mixed and interleaved I fear 😨 
I guess we can proceed iteratively. It doesn't hurt anyone for now.
